lib/net/ssh/connection/session.rb

```diff
@@ -205,11 +205,28 @@ def loop(wait=nil, &block)
     def process(wait=nil, &block)
       return false unless preprocess(&block)
 
-      r = listeners.keys
-      w = r.select { |w2| w2.respond_to?(:pending_write?) && w2.pending_write? }
-      readers, writers, = Net::SSH::Compat.io_select(r, w, nil, io_select_wait(wait))
-
-      postprocess(readers, writers)
+      wait_interval = io_select_wait(wait)
+
+      timed_out = false
+      listeners.each do |l, _|
```

Comment:
the first for write then for read is a concern indeed, see my comment with the ruby issue linked.